File: android/core/designsystem/src/main/kotlin/com/riskdetectedan/core/designsystem/isg/NovaChoiceField.kt
package com.riskdetectedan.core.designsystem.isg

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.scale
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalView
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.*
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.unit.sp
import androidx.compose.ui.unit.dp
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

/** The colour family of a choice: its tile, its level bars and the dot next to the chosen value (iOS `NovaChoiceTone`). */
enum class NovaChoiceTone {
    Neutral, Success, Warning, Danger;

    @Composable fun ink(): Color = when (this) {
        Neutral -> NovaColorToken.accentInk.color()
        Success -> NovaColorToken.statusSuccessInk.color()
        Warning -> NovaColorToken.statusWarningInk.color()
        Danger -> NovaColorToken.statusDangerInk.color()
    }

    @Composable fun soft(): Color = when (this) {
        Neutral -> NovaColorToken.surfaceMuted.color()
        Success -> NovaColorToken.statusSuccessBg.color()
        Warning -> NovaColorToken.statusWarningBg.color()
        Danger -> NovaColorToken.statusDangerBg.color()
    }
}

/**
 * One choice in a [NovaChoiceField] sheet (iOS `NovaChoiceOption`). [level] 1..3 draws rising level bars in the
 * tile (for ordered choices such as hazard classes); null draws [symbol].
 */
data class NovaChoiceOption<T>(
    val value: T,
    val title: String,
    val detail: String? = null,
    val tone: NovaChoiceTone = NovaChoiceTone.Neutral,
    val level: Int? = null,
    val symbol: String? = null,
)

/**
 * A form row that opens a bottom sheet of large, described choices (iOS `NovaChoiceField`). Nothing is chosen
 * until the user picks one: the row shows [placeholder], and the sheet closes on its own right after a pick.
 * Long lists get a search field and open full height; an optional field offers [noneTitle] as its first row
 * and reports it as null. [identifier] tags the row, "<identifier>.option.<index>", ".none" and ".search".
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun <T> NovaChoiceField(title: String, placeholder: String, symbol: String, options: List<NovaChoiceOption<T>>,
                        selection: T?, onSelect: (T?) -> Unit, identifier: String, modifier: Modifier = Modifier,
                        message: String? = null, enabled: Boolean = true, noneTitle: String? = null,
                        searchable: Boolean = options.size > 8,
                        /** A bordered field box (like [NovaChooserButton]) for forms of boxes; otherwise a plain row for a card with dividers. */
                        boxed: Boolean = false,
                        /** The sheet's heading when it should differ from [placeholder] (e.g. a placeholder that reads as a value). */
                        heading: String? = null,
                        /** Opens the sheet from outside, e.g. right after the step that needs this answer; reset through [onOpened]. */
                        openRequest: Boolean = false, onOpened: () -> Unit = {}) {
    var choosing by remember { mutableStateOf(false) }
    LaunchedEffect(openRequest) { if (openRequest && enabled) { choosing = true; onOpened() } }
    val chosen = options.firstOrNull { it.value == selection }
    val box = RoundedCornerShape(14.dp)
    Row(modifier.fillMaxWidth().heightIn(min = if (boxed) 52.dp else 48.dp).alpha(if (enabled) 1f else 0.5f)
        .then(if (boxed) Modifier.clip(box).novaControlBackground(14.dp).border(1.dp, NovaColorToken.border.color(), box) else Modifier)
        .novaRowPress(enabled = enabled) { choosing = true }.testTag(identifier)
        .semantics(mergeDescendants = true) { stateDescription = chosen?.title ?: noneTitle ?: placeholder }
        .then(if (boxed) Modifier.padding(horizontal = 12.dp) else Modifier),
        verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(10.dp)) {
        Box(Modifier.width(22.dp), contentAlignment = Alignment.Center) { NovaIcon(symbol, 17.dp) }
        Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(2.dp)) {
            if (chosen != null) {
                NovaText(title, style = NovaTypeToken.metaQuiet, color = NovaColorToken.textMuted.color())
                Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(7.dp)) {
                    if (chosen.tone != NovaChoiceTone.Neutral || chosen.level != null) Box(Modifier.size(8.dp).background(chosen.tone.ink(), CircleShape))
                    NovaText(chosen.title, style = NovaTypeToken.body, maxLines = 2)
                }
            } else if (noneTitle != null) {
                NovaText(title, style = NovaTypeToken.metaQuiet, color = NovaColorToken.textMuted.color())
                NovaText(noneTitle, style = NovaTypeToken.body, color = NovaColorToken.textSecondary.color())
            } else {
                NovaText(placeholder, style = NovaTypeToken.body, color = NovaColorToken.textPlaceholder.color())
            }
        }
        Box(Modifier.size(26.dp).background(NovaColorToken.surfaceMuted.color(), CircleShape).clearAndSetSemantics {},
            contentAlignment = Alignment.Center) {
            NovaIcon("chevron.down", 11.dp, tint = NovaColorToken.textMuted.color())
        }
    }
    if (choosing) {
        val sheet = rememberModalBottomSheetState(skipPartiallyExpanded = true)
        val scope = rememberCoroutineScope()
        val view = LocalView.current
        var query by remember { mutableStateOf("") }
        fun pick(value: T?) {
            onSelect(value)
            NovaHaptics.selection(view)
            // Long enough to see the tick land, short enough to feel instant.
            scope.launch { delay(220); sheet.hide() }.invokeOnCompletion { choosing = false }
        }
        val needle = query.trim()
        val found = options.withIndex().filter { (_, option) ->
            needle.isEmpty() || option.title.contains(needle, ignoreCase = true) || option.detail?.contains(needle, ignoreCase = true) == true
        }
        ModalBottomSheet(onDismissRequest = { choosing = false }, sheetState = sheet,
            containerColor = NovaColorToken.canvasSheet.color()) {
            Column(Modifier.fillMaxWidth().then(if (searchable) Modifier.fillMaxHeight(0.92f) else Modifier)
                .padding(start = 20.dp, end = 20.dp, top = 8.dp),
                verticalArrangement = Arrangement.spacedBy(14.dp)) {
                Column(verticalArrangement = Arrangement.spacedBy(6.dp)) {
                    // A field's name may serve as the heading; its required mark does not belong there.
                    NovaText((heading ?: placeholder).removeSuffix(" *"), style = NovaTypeToken.sheetTitle, modifier = Modifier.semantics { heading() })
                    if (message != null) NovaText(message, style = NovaTypeToken.metaQuiet, color = NovaColorToken.textSecondary.color())
                }
                if (searchable) ChoiceSearch(query, { query = it }, "$identifier.search")
                val none: @Composable () -> Unit = {
                    if (noneTitle != null && needle.isEmpty()) {
                        ChoiceCard(NovaChoiceOption(null, noneTitle), picked = selection == null, identifier = "$identifier.none") { pick(null) }
                    }
                }
                val empty: @Composable () -> Unit = {
                    if (found.isEmpty()) NovaText("Sonuç bulunamadı", Modifier.fillMaxWidth().padding(vertical = 24.dp),
                        NovaTypeToken.metaQuiet, NovaColorToken.textSecondary.color(), textAlign = TextAlign.Center)
                }
                // Long lists (people, workplaces) draw only the rows on screen.
                if (searchable) LazyColumn(Modifier.weight(1f), contentPadding = PaddingValues(bottom = 28.dp),
                    verticalArrangement = Arrangement.spacedBy(10.dp)) {
                    item { none() }
                    items(found, key = { it.index }) { (index, option) ->
                        ChoiceCard(option, picked = option.value == selection, identifier = "$identifier.option.$index") { pick(option.value) }
                    }
                    item { empty() }
                } else Column(Modifier.weight(1f, fill = false).verticalScroll(rememberScrollState()).padding(bottom = 28.dp),
                    verticalArrangement = Arrangement.spacedBy(10.dp)) {
                    none()
                    found.forEach { (index, option) ->
                        ChoiceCard(option, picked = option.value == selection, identifier = "$identifier.option.$index") { pick(option.value) }
                    }
                    empty()
                }
            }
        }
    }
}

/** The sheet's search box (the same look as [NovaChooserPanel]'s). */
@Composable
private fun ChoiceSearch(query: String, onChange: (String) -> Unit, identifier: String) {
    Row(Modifier.fillMaxWidth().heightIn(min = 46.dp).background(NovaColorToken.surface.color(), RoundedCornerShape(14.dp))
        .border(1.dp, NovaColorToken.borderMuted.color(), RoundedCornerShape(14.dp)).padding(horizontal = 14.dp),
        verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(8.dp)) {
        NovaIcon("magnifyingglass", 15.dp, tint = NovaColorToken.textMuted.color())
        Box(Modifier.weight(1f)) {
            if (query.isEmpty()) NovaText("Ara", style = NovaTypeToken.body, color = NovaColorToken.textPlaceholder.color())
            BasicTextField(query, onChange, Modifier.fillMaxWidth().testTag(identifier).semantics { contentDescription = "Seçeneklerde ara" },
                singleLine = true, textStyle = TextStyle(fontFamily = NovaFontFamilyPublic, fontSize = 14.sp, color = NovaColorToken.text.color()),
                cursorBrush = SolidColor(NovaColorToken.text.color()))
        }
        if (query.isNotEmpty()) Box(Modifier.size(32.dp).novaPress(onClickLabel = "Aramayı temizle") { onChange("") },
            contentAlignment = Alignment.Center) { NovaIcon("xmark", 14.dp) }
    }
}

/** A described choice is a large card with a tile; a plain one (a workplace, a person) is a compact row with the same tick. */
@Composable
private fun <T> ChoiceCard(option: NovaChoiceOption<T>, picked: Boolean, identifier: String, onClick: () -> Unit) {
    val ink = option.tone.ink()
    val shape = RoundedCornerShape(18.dp)
    val tick by animateFloatAsState(if (picked) 1f else 0f, label = "choice.tick")
    val hasTile = option.level != null || option.symbol != null
    val plain = !hasTile && option.detail == null
    val pickedBackground = if (option.tone == NovaChoiceTone.Neutral) NovaColorToken.accentSoft.color() else option.tone.soft().copy(alpha = 0.55f)
    Row(Modifier.fillMaxWidth().heightIn(min = if (plain) 52.dp else 0.dp).clip(shape)
        .background(if (picked) pickedBackground else NovaColorToken.surface.color(), shape)
        .border(if (picked) 1.5.dp else 1.dp, if (picked) ink else NovaColorToken.borderMuted.color(), shape)
        .novaRowPress(role = Role.RadioButton, onClick = onClick).testTag(identifier)
        .semantics(mergeDescendants = true) { selected = picked }
        .padding(horizontal = 14.dp, vertical = if (plain) 12.dp else 14.dp),
        verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(14.dp)) {
        if (hasTile) Box(Modifier.size(46.dp).background(option.tone.soft(), RoundedCornerShape(13.dp)).clearAndSetSemantics {},
            contentAlignment = Alignment.Center) {
            val level = option.level
            if (level != null) Row(verticalAlignment = Alignment.Bottom, horizontalArrangement = Arrangement.spacedBy(3.dp)) {
                (1..3).forEach { bar ->
                    Box(Modifier.width(5.dp).height((6 + bar * 5).dp)
                        .background(if (bar <= level) ink else ink.copy(alpha = 0.22f), CircleShape))
                }
            } else option.symbol?.let { NovaIcon(it, 18.dp, tint = ink) }
        }
        Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(3.dp)) {
            NovaText(option.title, style = if (plain) NovaTypeToken.body else NovaTypeToken.cardTitle)
            option.detail?.let { NovaText(it, style = NovaTypeToken.metaQuiet, color = NovaColorToken.textSecondary.color()) }
        }
        Box(Modifier.size(24.dp).clearAndSetSemantics {}, contentAlignment = Alignment.Center) {
            Box(Modifier.matchParentSize().border(1.5.dp, NovaColorToken.border.color(), CircleShape))
            Box(Modifier.matchParentSize().scale(tick).background(ink, CircleShape), contentAlignment = Alignment.Center) {
                NovaIcon("checkmark", 11.dp, tint = Color.White)
            }
        }
    }
}

/** Hazard classes as described choices (iOS `NovaHazardChoice`), keyed by the stored ids "low", "medium", "high". */
object NovaHazardChoice {
    const val placeholder = "Tehlike sınıfı seçin"
    const val message = "Resmî tehlike sınıfı, işyerinin NACE koduna göre belirlenir."
    val options: List<NovaChoiceOption<String>> = listOf(
        NovaChoiceOption("low", "Az Tehlikeli", "Risk değerlendirmesi 6 yılda bir yenilenir · en az C sınıfı uzman", NovaChoiceTone.Success, level = 1),
        NovaChoiceOption("medium", "Tehlikeli", "Risk değerlendirmesi 4 yılda bir yenilenir · en az B sınıfı uzman", NovaChoiceTone.Warning, level = 2),
        NovaChoiceOption("high", "Çok Tehlikeli", "Risk değerlendirmesi 2 yılda bir yenilenir · A sınıfı uzman", NovaChoiceTone.Danger, level = 3),
    )
}

File: App/DesignSystem/ISG/NovaChoiceField.swift
import SwiftUI

/// The colour family of a choice: its tile, its level bars and the dot next
/// to the chosen value.
enum NovaChoiceTone {
    case neutral, success, warning, danger

    func ink(in scheme: ColorScheme) -> Color {
        switch self {
        case .neutral: return NovaColorToken.accentInk.color(in: scheme)
        case .success: return NovaColorToken.statusSuccessInk.color(in: scheme)
        case .warning: return NovaColorToken.statusWarningInk.color(in: scheme)
        case .danger: return NovaColorToken.statusDangerInk.color(in: scheme)
        }
    }

    func soft(in scheme: ColorScheme) -> Color {
        switch self {
        case .neutral: return NovaColorToken.surfaceMuted.color(in: scheme)
        case .success: return NovaColorToken.statusSuccessBg.color(in: scheme)
        case .warning: return NovaColorToken.statusWarningBg.color(in: scheme)
        case .danger: return NovaColorToken.statusDangerBg.color(in: scheme)
        }
    }
}

/// One choice in a `NovaChoiceField` sheet.
struct NovaChoiceOption<Value: Hashable>: Identifiable {
    let value: Value
    let title: String
    var detail: String? = nil
    var tone: NovaChoiceTone = .neutral
    /// 1...3 draws rising level bars in the tile (for ordered choices such as
    /// hazard classes); nil draws `symbol`.
    var level: Int? = nil
    var symbol: String? = nil
    var id: Value { value }
}

/// A form row that opens a bottom sheet of large, described choices. Nothing is
/// chosen until the user picks one: the row shows the placeholder, and the
/// sheet closes on its own right after a pick. Long lists get a search field
/// and open full height; an optional field offers `noneTitle` as its first row.
struct NovaChoiceField<Value: Hashable>: View {
    /// The row's name, shown above the chosen value ("Tehlike sınıfı *").
    let title: String
    /// Shown in the row while nothing is chosen, and as the sheet's heading.
    let placeholder: String
    let symbol: String
    var message: String? = nil
    let options: [NovaChoiceOption<Value>]
    @Binding var selection: Value?
    /// Accessibility identifier stem: the row, "<stem>.option.<index>",
    /// "<stem>.none" and "<stem>.search".
    var identifier: String
    /// For an optional field: the row that clears the choice ("Seçilmedi").
    var noneTitle: String? = nil
    /// nil: a search field once the list is longer than eight.
    var searchable: Bool? = nil
    /// A bordered field box, for forms whose other fields are boxes too;
    /// otherwise a plain row for a card with dividers.
    var boxed = false
    /// The sheet's heading when it should differ from the placeholder (e.g. a
    /// placeholder that reads as a value, "Firma geneli").
    var heading: String? = nil
    /// Opens the sheet from outside, e.g. right after the step that needs
    /// this answer; set back to false once the sheet is up.
    var openRequest: Binding<Bool>? = nil
    @State private var choosing = false
    @State private var query = ""
    @Environment(\.colorScheme) private var scheme
    @Environment(\.isEnabled) private var enabled

    private var chosen: NovaChoiceOption<Value>? { options.first { $0.value == selection } }
    private var searches: Bool { searchable ?? (options.count > 8) }
    private var matches: [(offset: Int, element: NovaChoiceOption<Value>)] {
        let needle = query.trimmingCharacters(in: .whitespacesAndNewlines)
        return options.enumerated().filter { needle.isEmpty || $0.element.title.localizedStandardContains(needle)
            || ($0.element.detail?.localizedStandardContains(needle) ?? false) }
    }

    var body: some View {
        Button { choosing = true } label: {
            HStack(spacing: 10) {
                NovaIcon(symbol: symbol, size: 17).frame(width: 22)
                if let chosen {
                    VStack(alignment: .leading, spacing: 2) {
                        NovaText(text: title, style: .metaQuiet, color: NovaColorToken.textMuted.color(in: scheme))
                        HStack(spacing: 7) {
                            if chosen.tone != .neutral || chosen.level != nil {
                                Circle().fill(chosen.tone.ink(in: scheme)).frame(width: 8, height: 8)
                            }
                            NovaText(text: chosen.title, style: .body).lineLimit(2)
                        }
                    }
                } else if let noneTitle {
                    VStack(alignment: .leading, spacing: 2) {
                        NovaText(text: title, style: .metaQuiet, color: NovaColorToken.textMuted.color(in: scheme))
                        NovaText(text: noneTitle, style: .body, color: NovaColorToken.textSecondary.color(in: scheme))
                    }
                } else {
                    NovaText(text: placeholder, style: .body, color: NovaColorToken.textPlaceholder.color(in: scheme))
                }
                Spacer(minLength: 8)
                Image(systemName: "chevron.down")
                    .font(.system(size: 11, weight: .bold))
                    .foregroundStyle(NovaColorToken.textMuted.color(in: scheme))
                    .frame(width: 26, height: 26)
                    .background(NovaColorToken.surfaceMuted.color(in: scheme), in: Circle())
            }
            .frame(maxWidth: .infinity, minHeight: 48, alignment: .leading)
            .padding(.horizontal, boxed ? 12 : 0).padding(.vertical, boxed ? 2 : 0)
            .modifier(NovaChoiceBox(active: boxed))
            .contentShape(Rectangle())
        }
        .buttonStyle(NovaRowPressStyle())
        .opacity(enabled ? 1 : 0.5)
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(Text(title))
        .accessibilityValue(Text(chosen?.title ?? noneTitle ?? placeholder))
        .accessibilityAddTraits(.isButton)
        .accessibilityIdentifier(identifier)
        .onChange(of: openRequest?.wrappedValue ?? false) { open in
            guard open, enabled else { return }
            choosing = true
            openRequest?.wrappedValue = false
        }
        .sheet(isPresented: $choosing, onDismiss: { query = "" }) {
            if searches {
                VStack(alignment: .leading, spacing: 14) {
                    sheetHeading
                    searchField
                    ScrollView { list.padding(.bottom, 20) }
                        .scrollDismissesKeyboard(.interactively)
                }
                .padding(.horizontal, 20).padding(.top, 28)
                .presentationDetents([.large])
                .presentationDragIndicator(.visible)
            } else {
                RDContentSizedSheet {
                    VStack(alignment: .leading, spacing: 18) { sheetHeading; list }
                        .padding(.horizontal, 20).padding(.top, 28).padding(.bottom, 20)
                } footer: {
                    EmptyView()
                }
                .presentationDragIndicator(.visible)
            }
        }
    }

    private var sheetHeading: some View {
        VStack(alignment: .leading, spacing: 6) {
            // A field's name may serve as the heading; its required mark does not belong there.
            NovaText(text: (heading ?? placeholder).replacingOccurrences(of: " *", with: ""), style: .sheetTitle).accessibilityAddTraits(.isHeader)
            if let message {
                NovaText(text: message, style: .metaQuiet, color: NovaColorToken.textSecondary.color(in: scheme))
                    .fixedSize(horizontal: false, vertical: true)
            }
        }
    }

    private var searchField: some View {
        HStack(spacing: 8) {
            Image(systemName: "magnifyingglass").foregroundStyle(NovaColorToken.textMuted.color(in: scheme))
            TextField(NovaChoiceText.search, text: $query)
                .font(NovaFont.font(.body)).autocorrectionDisabled().submitLabel(.search)
                .accessibilityIdentifier("\(identifier).search")
        }
        .padding(.horizontal, 14).frame(minHeight: 46)
        .background(NovaColorToken.surface.color(in: scheme), in: RoundedRectangle(cornerRadius: 14, style: .continuous))
        .overlay(RoundedRectangle(cornerRadius: 14, style: .continuous).strokeBorder(NovaColorToken.borderMuted.color(in: scheme), lineWidth: 1))
    }

    private var list: some View {
        let found = matches
        let needle = query.trimmingCharacters(in: .whitespacesAndNewlines)
        // Long lists (people, workplaces) draw only the rows on screen.
        return NovaChoiceStack(lazy: searches) {
            if let noneTitle, needle.isEmpty {
                card(title: noneTitle, detail: nil, tone: .neutral, tile: nil, picked: selection == nil) { pick(nil) }
                    .accessibilityIdentifier("\(identifier).none")
            }
            ForEach(found, id: \.element.id) { index, option in
                card(title: option.title, detail: option.detail, tone: option.tone,
                     tile: option.level != nil || option.symbol != nil ? option : nil,
                     picked: option.value == selection) { pick(option.value) }
                    .accessibilityIdentifier("\(identifier).option.\(index)")
            }
            if found.isEmpty {
                NovaText(text: NovaChoiceText.empty,
                    style: .metaQuiet, color: NovaColorToken.textSecondary.color(in: scheme))
                    .frame(maxWidth: .infinity).padding(.vertical, 24)
            }
        }
    }

    private func pick(_ value: Value?) {
        selection = value
        NovaHaptics.selection()
        // Long enough to see the tick land, short enough to feel instant.
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.22) { choosing = false }
    }

    /// A described choice is a large card with a tile; a plain one (a
    /// workplace, a person) is a compact row with the same tick.
    private func card(title: String, detail: String?, tone: NovaChoiceTone, tile option: NovaChoiceOption<Value>?,
                      picked: Bool, action: @escaping () -> Void) -> some View {
        let ink = tone.ink(in: scheme)
        let plain = option == nil && detail == nil
        return Button(action: action) {
            HStack(spacing: 14) {
                if let option { tile(option) }
                VStack(alignment: .leading, spacing: 3) {
                    NovaText(text: title, style: plain ? .body : .cardTitle)
                    if let detail {
                        NovaText(text: detail, style: .metaQuiet, color: NovaColorToken.textSecondary.color(in: scheme))
                            .fixedSize(horizontal: false, vertical: true)
                    }
                }
                Spacer(minLength: 8)
                ZStack {
                    Circle().strokeBorder(picked ? ink : NovaColorToken.border.color(in: scheme), lineWidth: picked ? 0 : 1.5)
                    if picked {
                        Circle().fill(ink)
                        Image(systemName: "checkmark").font(.system(size: 11, weight: .heavy)).foregroundStyle(.white)
                    }
                }
                .frame(width: 24, height: 24)
                .animation(.spring(response: 0.25, dampingFraction: 0.7), value: picked)
            }
            .padding(.horizontal, 14).padding(.vertical, plain ? 12 : 14)
            .frame(maxWidth: .infinity, minHeight: plain ? 52 : 0, alignment: .leading)
            .background(picked ? (tone == .neutral ? NovaColorToken.accentSoft.color(in: scheme) : tone.soft(in: scheme).opacity(0.55))
                : NovaColorToken.surface.color(in: scheme),
                in: RoundedRectangle(cornerRadius: 18, style: .continuous))
            .overlay(RoundedRectangle(cornerRadius: 18, style: .continuous)
                .strokeBorder(picked ? ink : NovaColorToken.borderMuted.color(in: scheme), lineWidth: picked ? 1.5 : 1))
            .contentShape(RoundedRectangle(cornerRadius: 18, style: .continuous))
        }
        .buttonStyle(NovaRowPressStyle())
        .accessibilityElement(children: .combine)
        .accessibilityAddTraits(picked ? [.isButton, .isSelected] : .isButton)
    }

    private func tile(_ option: NovaChoiceOption<Value>) -> some View {
        let ink = option.tone.ink(in: scheme)
        return ZStack {
            RoundedRectangle(cornerRadius: 13, style: .continuous).fill(option.tone.soft(in: scheme))
            if let level = option.level {
                HStack(alignment: .bottom, spacing: 3) {
                    ForEach(1...3, id: \.self) { bar in
                        Capsule().fill(bar <= level ? ink : ink.opacity(0.22))
                            .frame(width: 5, height: CGFloat(6 + bar * 5))
                    }
                }
            } else if let symbol = option.symbol {
                NovaIcon(symbol: symbol, size: 18).foregroundStyle(ink)
            }
        }
        .frame(width: 46, height: 46)
        .accessibilityHidden(true)
    }
}

/// A column that builds its rows lazily when the list can be long.
private struct NovaChoiceStack<Content: View>: View {
    let lazy: Bool
    @ViewBuilder let content: () -> Content
    var body: some View {
        if lazy { LazyVStack(spacing: 10, content: content) } else { VStack(spacing: 10, content: content) }
    }
}

/// The chooser's own words.
enum NovaChoiceText {
    /// The row that clears an optional choice.
    static var none: String { RDLocalization.string("localizable.nova.choice.none", table: .localizable, fallback: "Seçilmedi") }
    static var search: String { RDLocalization.string("localizable.nova.choice.search", table: .localizable, fallback: "Ara") }
    static var empty: String { RDLocalization.string("localizable.nova.choice.empty", table: .localizable, fallback: "Sonuç bulunamadı") }
}

/// The field box other form controls use, when the chooser sits among them.
private struct NovaChoiceBox: ViewModifier {
    let active: Bool
    @ViewBuilder func body(content: Content) -> some View {
        if active { content.novaControlBackground(cornerRadius: 14) } else { content }
    }
}

/// Hazard classes as described choices: what each class means for the expert's
/// daily work (renewal period of the risk assessment, minimum expert class).
enum NovaHazardChoice {
    static var placeholder: String {
        RDLocalization.string("localizable.nova.hazard.choice.placeholder", table: .localizable, fallback: "Tehlike sınıfı seçin")
    }
    static var message: String {
        RDLocalization.string("localizable.nova.hazard.choice.message", table: .localizable,
            fallback: "Resmî tehlike sınıfı, işyerinin NACE koduna göre belirlenir.")
    }
    static var options: [NovaChoiceOption<CompanyHazardClass>] {
        [.init(value: .low, title: CompanyHazardClass.low.title,
               detail: RDLocalization.string("localizable.nova.hazard.choice.low", table: .localizable,
                   fallback: "Risk değerlendirmesi 6 yılda bir yenilenir · en az C sınıfı uzman"), tone: .success, level: 1),
         .init(value: .medium, title: CompanyHazardClass.medium.title,
               detail: RDLocalization.string("localizable.nova.hazard.choice.medium", table: .localizable,
                   fallback: "Risk değerlendirmesi 4 yılda bir yenilenir · en az B sınıfı uzman"), tone: .warning, level: 2),
         .init(value: .high, title: CompanyHazardClass.high.title,
               detail: RDLocalization.string("localizable.nova.hazard.choice.high", table: .localizable,
                   fallback: "Risk değerlendirmesi 2 yılda bir yenilenir · A sınıfı uzman"), tone: .danger, level: 3)]
    }
}
